### PR TITLE
Add an ETW event for reporting richer debug information back

### PR DESCRIPTION
**Pull request by jakobbotsch**
This adds some richer debug information in the form of inline trees and IL<->native mappings with inlinee information, plus an ETW event in the private runtime provider to report it back. A COMPlus variable is used to opt-in to generating and storing this larger debug information.

On the JIT side, we now store a few more fields of data for each inline context in release builds:
* Its ordinal which is required to be incrementally assigned
* The actual IL offset of the IL instruction leading to the creation of the inline context
* The method handle of the inlinee

On the EE side we store the new debug information together with the normal debug info, reusing the flag byte used for patchpoint information.

The hope is that the new richer format of debug info can eventually replace the old one, but this won't happen in .NET 7 as I expect it will take some time to come up with the right format for this data and the right set of fields to expose. In the meantime this environment variable and event/debugging APIs are mainly to be used as an opt-in way to start incrementally prototyping on the tooling side without needing special checked builds of the runtime/JIT.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
I will work a bit on revising this to also (when opted-in) store the rich debug info so that rundown can work too.

File: src/coreclr/vm/debuginfostore.cpp [hunk 1]
```diff
@@ -680,26 +701,77 @@ PatchpointInfo * CompressDebugInfo::RestorePatchpointInfo(IN PTR_BYTE pDebugInfo
...
-
     // Check flag byte.
     BYTE flagByte = *pDebugInfo;
     pDebugInfo++;
 
-    if (flagByte == 1)
+    if ((flagByte & EXTRA_DEBUG_INFO_PATCHPOINT) == 0)
+        return NULL;
+
+    return static_cast<PatchpointInfo*>(PTR_READ(dac_cast<TADDR>(pDebugInfo), dac_cast<PTR_PatchpointInfo>(pDebugInfo)->PatchpointInfoSize()));
```
Comment (jakobbotsch) [inline src/coreclr/vm/debuginfostore.cpp:711]: This struct has a flexible array at the end, IIUC the previous code was not correct in DAC builds.

File: src/coreclr/vm/debuginfostore.cpp [hunk 2]
```diff
@@ -711,33 +783,43 @@ void CompressDebugInfo::EnumMemoryRegions(CLRDataEnumMemoryFlags flags, PTR_BYTE
...
 
     NibbleReader r(pDebugInfo, 12 /* maximum size of compressed 2 UINT32s */);
 
     ULONG cbBounds = r.ReadEncodedU32();
     ULONG cbVars   = r.ReadEncodedU32();
 
-    DacEnumMemoryRegion(dac_cast<TADDR>(pDebugInfo), r.GetNextByteIndex() + cbBounds + cbVars);
+    pDebugInfo += r.GetNextByteIndex() + cbBounds + cbVars;
+
+    DacEnumMemoryRegion(dac_cast<TADDR>(pStart), pDebugInfo - pStart);
```
Comment (jakobbotsch) [inline src/coreclr/vm/debuginfostore.cpp:822]: This code also seemed wrong before -- it would not include all the debug information in some cases.

**Comment by jakobbotsch** [comment]
Failure looks like #71243.

Still have a few things to discuss with @brianrob, but I think this should be ready for a review of the JIT and VM parts, PTAL @AndyAyersMS and @jkotas.

**Comment by jkotas** [comment]
> The hope is that the new richer format of debug info can eventually replace the old one, but this won't happen in .NET 7 as I expect it will take some time to come up with the right format for this data and the right set of fields to expose.

It does not look right to me that the format contains method handles. It makes it impossible to serialize it into AOT compiled images, and it makes it expensive to consume it as events - the consumer needs to keep track of the method handles.

Is it really worth it go with this temporary format?

**Comment by jakobbotsch** [comment]
> It does not look right to me that the format contains method handles. It makes it impossible to serialize it into AOT compiled images, and it makes it expensive to consume it as events - the consumer needs to keep track of the method handles.

The temporary format I am referring to here is how we compress it and store it in the VM. For `InlineTreeNode` using the method handle to communicate back from the JIT seems appropriate to me.

I don't expect the final format that we will be serializing to to contain methods handles, but for the ETW events the method handle is the easiest way to cross reference today (it is what MethodILToNativeMap, MethodLoadUnload and MethodDetails are using), so I expect the VM will always need to be able to restore to the format containing method handles for inlinees.

We have some leeway in how we serialize it and store it in the VM and NAOT images, these don't necessarily have to match. However, in the end I expect the IL offset in the inline tree to be enough to reconstruct the exact handle in almost all cases, except for a few rare cases where we might just give up or do something more expensive. For example it could be some kind of fixup-like construct in AOT'd images, while it can be just a method handle in the VM's compressed info, somewhat like `ELEMENT_TYPE_INTERNAL`.

**Review by davidwrighton — APPROVED** [review]
So this looks ok. The detail around needing to process MethodHandles for the ETW event processing is terribly inconvenient, and unfortunate, but I ended up with the same general restrictions when working with PGO data. You can either have precise data, which will likely involve working with a MethodHandle and all of the difficulties presented by that, or you end up generating giant strings, which aren't really parseable. For native AOT debugging, I suspect that representation as giant strings, will be what gets injected into the actual native debug data.

**Comment by davidwrighton** [comment]
@jkotas could you express an opinion here in response to my statement? I'm ok with missing the Preview snap for this, as nobody is really expecting to take a strong dependency on this I expect.

**Comment by jkotas** [comment]
Sounds reasonable to me. Thanks for sharing the broader context.

File: src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp [hunk 3]
```diff
@@ -1167,6 +1167,14 @@ void interceptor_ICJI::setVars(CORINFO_METHOD_HANDLE         ftn,   // [IN] meth
     original_ICorJitInfo->setVars(ftn, cVars, vars);
 }
 
+void interceptor_ICJI::reportRichMappings(ICorDebugInfo::InlineTreeNode*    inlineTreeNodes,
+                                          uint32_t                          numInlineTreeNodes,
+                                          ICorDebugInfo::RichOffsetMapping* mappings,
+                                          uint32_t                          numMappings)
+{
+    mc->cr->AddCall("reportRichMappings");
```
Comment (AndyAyersMS) [inline src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp:1175]: Do you plan on eventually capturing this info? Seem fine to defer for now, but maybe add a TODO comment.

File: src/coreclr/vm/codeman.cpp [hunk 4]
```diff
@@ -6029,6 +6074,17 @@ BOOL ReadyToRunJitManager::GetBoundariesAndVars(
...
 
+BOOL ReadyToRunJitManager::GetRichDebugInfo(
+    const DebugInfoRequest& request,
+    IN FP_IDS_NEW fpNew, IN void* pNewData,
+    OUT ICorDebugInfo::InlineTreeNode** ppInlineTree,
+    OUT ULONG32* pNumInlineTree,
+    OUT ICorDebugInfo::RichOffsetMapping** ppRichMappings,
+    OUT ULONG32* pNumRichMappings)
+{
+    return FALSE;
```
Comment (AndyAyersMS) [inline src/coreclr/vm/codeman.cpp:6085]: Seems like eventually we will want to support rich debug and prejitting. 

Maybe another TODO?

**Review by AndyAyersMS — APPROVED** [review]
JIT and debug info store changes look good to me.

File: src/coreclr/vm/codeman.cpp (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/vm/codeman.cpp:6085]: Yes, I will work on this in .NET 8 as part of working out a better storage format for the data.

File: src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp:1175]: Ah, good point, I forgot about this. Yes, let me add a todo and do it in a follow-up.
More importantly we need to free the arrays here.

**Review by brianrob — APPROVED** [review]
LGTM from the tracing perspective.

**Pull request merged**